### PR TITLE
#300 Add .cs class for missing resource types in generator code

### LINKED ISSUE #268: Export KeyVault "child" resources and add support in Generator
`KeyVaultKey`, `KeyVaultCertificate`, and `KeyVaultSecret` are not present in the `ResourceType` enum even though there are `Confirm` functions supporting them. 

- [ ] Add those resources to the `ResourceType` enum
- [ ] Add examples of using `Confirm-AzBPResource` with those resources
- [ ] Add support for test generation in the Generator code

### LINKED ISSUE #300: Add .cs class for missing resource types in generator code
As an engineer, I would like the resource types currently missing a class in `BenchPress\Generators\ResourceTypes` folder to be added.

This is to keep both parts of the project in sync.

The bicep resource is Microsoft.DocumentDB/databaseAccounts/sqlRoleAssignments

- [x] Create `ApiManagement.cs`
- [x] Create `ApiManagementApi.cs`
- [x] Create `ApiManagementDiagnostic.cs`
- [x] Create `ApiManagementLogger.cs`
- [x] Create `ApiManagementPolicy.cs`
- [ ] Create `RoleAssignment.cs`
- [x] Create `ContainerApp.cs`

### PR DESCRIPTION
**Pull request by marclerwick**
Added missing resource types to the Generator code.

*NOTE* Due to limitations in the Generator code Key Vault Certificate and Role Assignment cannot be supported at this time. Specifically Key Vault Certificate cannot be created using bicep so the proper variable names cannot be determined for the Generator. Role Assignments have additional properties that are needed that cannot be derived from the generated ARM template.

Closes #300 
Closes #268

### CODE REVIEW

**Review by ozboms — APPROVED** [review]

File: BenchPress/Generators/ResourceTypes/ApiManagementApi.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public override string FunctionPrefix => "ApiManagementApi";
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "ApiManagementApi"),
+            Param("ResourceName", m.ResourceName),
+            Param("ResourceGroupName", m.ExtraProperties["resourceGroup"]),
+            Param("ServiceName", m.ExtraProperties["serviceName"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/ApiManagementApi.cs:20]: The type of the "parent" resource is Microsoft.ApiManagement/service. So shouldn't the key be `service`?

File: BenchPress/Generators/ResourceTypes/ApiManagementDiagnostic.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    public override string FunctionPrefix => "ApiManagementDiagnostic";
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "ApiManagementDiagnostic"),
+            Param("ResourceName", m.ResourceName),
+            Param("ResourceGroupName", m.ExtraProperties["resourceGroup"]),
+            Param("ServiceName", m.ExtraProperties["serviceName"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/ApiManagementDiagnostic.cs:20]: Same comment. Should key be `service`?

File: BenchPress/Generators/ResourceTypes/ApiManagementLogger.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    public override string FunctionPrefix => "ApiManagementLogger";
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "ApiManagementLogger"),
+            Param("ResourceName", m.ResourceName),
+            Param("ResourceGroupName", m.ExtraProperties["resourceGroup"]),
+            Param("ServiceName", m.ExtraProperties["serviceName"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/ApiManagementLogger.cs:20]: Same comment. Should key be `service`?

File: BenchPress/Generators/ResourceTypes/ApiManagementPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "ApiManagementPolicy"),
+            Param("ResourceName", m.ResourceName),
+            Param("ResourceGroupName", m.ExtraProperties["resourceGroup"]),
+            Param("ServiceName", m.ExtraProperties["serviceName"]),
+            Param("ApiId", m.ExtraProperties["apiName"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/ApiManagementPolicy.cs:21]: Similar comment. If "parent"/dependsOn resources are of type "Microsoft.ApiManagement/service/apis", should the keys be `service` and `apis`?

File: BenchPress/Generators/ResourceTypes/KeyVaultKey.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    public override string Prefix => "kvk";
+    public override string FunctionPrefix => "KeyVaultKey";
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "KeyVaultKey"),
+            Param("ResourceName", m.ResourceName),
+            Param("KeyVaultName", m.ExtraProperties["name"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: Similar comment. Should the key be `vaults`?

File: BenchPress/Generators/ResourceTypes/KeyVaultSecret.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    public override string Prefix => "kvs";
+    public override string FunctionPrefix => "KeyVaultSecret";
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "KeyVaultSecret"),
+            Param("ResourceName", m.ResourceName),
+            Param("KeyVaultName", m.ExtraProperties["name"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/KeyVaultSecret.cs:19]: Key should be `vaults`?

**Review by msha1026 — CHANGES_REQUESTED** [review]

**Comment by msha1026** [comment]
Some additional Qs, should follow up issues be created to investigate/address the stuff you noted in your PR description? Specifically, should testing KV certificate be a part of Benchpress if it's not possible to create through bicep? (Also asking this because I came across that issue retesting all of our examples 😆 )

Lastly, this work also closes #268.

File: BenchPress/Generators/ResourceTypes/ApiManagementApi.cs (hunk 1 above)
Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/ApiManagementApi.cs:20]: It's `serviceName` in the generated ARM template.

File: BenchPress/Generators/ResourceTypes/ApiManagementPolicy.cs (hunk 4 above)
Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/ApiManagementPolicy.cs:21]: Same, it's `apiName` in the generated ARM template.

File: BenchPress/Generators/ResourceTypes/KeyVaultKey.cs (hunk 5 above)
Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: Maybe I misunderstood our conversation the other day, but all of the `ExtraProperties` keys come from the generated ARM template, no?

Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: Example:

``` JSON
    {
      "type": "Microsoft.KeyVault/vaults/keys",
      "apiVersion": "2022-07-01",
      "name": "[format('{0}/{1}', parameters('name'), 'samplekey')]",
      "properties": {
        "kty": "RSA",
        "keyOps": [],
        "keySize": 2048,
        "curveName": "P-521"
      },
      "dependsOn": [
        "[resourceId('Microsoft.KeyVault/vaults', parameters('name'))]"
      ]
    }

```

**Comment by marclerwick** [comment]
> Some additional Qs, should follow up issues be created to investigate/address the stuff you noted in your PR description? Specifically, should testing KV certificate be a part of Benchpress if it's not possible to create through bicep? (Also asking this because I came across that issue retesting all of our examples 😆 )
> 
> Lastly, this work also closes #268.

I indeed was planning on adding a work item for Role Assignments.

Regarding KV Certificate, specifically as it pertains to the Generator code, if bicep cannot create it, the Generator cannot generate it. In the context of the vision for Benchpress (as a general Azure IaC testing framework) KV Certificate should be supported for testing.

File: BenchPress/Generators/ResourceTypes/KeyVaultKey.cs (hunk 5 above)
Comment (msha1026) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: Yes, it does come from the ARM template. I explained a bit how the extraProperties dictionary is created in the comments of the method: https://github.com/Azure/benchpress/blob/1f759817506f678202ee7d5300ef7c8d6502c281/BenchPress/Generators/AzureDeploymentImporter.cs#L145

For your example, the `extraProperties` dictionary would look like the following:
```
{
    "vaults": "parameters('name'))]" <- This reminds me that I need to create an issue to correctly strip any trailing braces
}
```

For something like ApiManagementPolicy, I assume the `dependsOn` object looks like this:
```
"[resourceId('Microsoft.ApiManagement/service/apis', parameters('servicename'), parameters('apiname'))]"
```
which would result in the following dictionary:
```
{
    "service": parameters('servicename')
    "apis": parameters('apiname'))]
}
```

In short, the resource type is used to create the keys, and then the parameter names are matched with the respective keys.

Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: So, did I do it right? LOL 

The generated ARM is:

``` JSON
    {
      "type": "Microsoft.ApiManagement/service/apis/policies",
      "apiVersion": "2022-08-01",
      "name": "[format('{0}/{1}/{2}', parameters('serviceName'), parameters('apiName'), parameters('policyName'))]",
      "properties": {
        "format": "rawxml",
        "value": "[variables('$fxv#0')]"
      },
      "dependsOn": [
        "[resourceId('Microsoft.ApiManagement/service/apis', parameters('serviceName'), parameters('apiName'))]"
      ]
    }
```

Comment (msha1026) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: No, you did not.

Again, if that is the generated ARM for API Policies, the following key/value pairs will be set in the `extraProperties` dictionary:
```
{
    "service": parameters('serviceName')
    "apis": parameters('apiName'))]
}
```
Then in the `ApiManagementApi` class, the following `Param`s can be initialized:
```
Param("ServiceName", m.ExtraProperties["service"]),
Param("ApiId", m.ExtraProperties["apis"])
```
However, if there is still confusion, I would prefer to understand anything else I can clear up. Or we can go over in design session, so that it's clear how to introduce new ResourceType classes.

Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: AH!!! I GET IT!!!

I don't know why it took me so long to wrap my head around it, but now it makes sense.

Update coming up!

Comment (msha1026) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: Yay, awesome! Apologies if I introduced any confusion. If you feel like more descriptive comments could be added to this method, https://github.com/Azure/benchpress/blob/1f759817506f678202ee7d5300ef7c8d6502c281/BenchPress/Generators/AzureDeploymentImporter.cs#L167
I am more than happy to add some better documentation and examples! It made so much sense to me during implementation. Looking back though, it's definitely super confusing LOL  🤣

Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: Where this really threw me for a loop is when there is more than one element in the path (e.g., `Microsoft.ApiManagement/service/apis`) and how that actually created two keys `service` and `apis` and how the parameters map to those two keys. Essentially, what is the relationship between the path, the keys, and the values.

Now that I understand that, it's pretty easy.

The only thing that still kind of makes me think twice is that I wonder if this is a universal truth. Meaning will this relationship between the path, the keys, and the values always be true (e.g., will there ever be a /service/someservice has three parameters). So far, I haven't found on, so I assume that it's a built in universal truth.

Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/KeyVaultKey.cs:19]: I see in the code that if the path's count doesn't match the parameters count it doesn't do anything! LOL

File: BenchPress/Generators/ResourceTypes/ApiManagementApi.cs (hunk 1 above)
Comment (marclerwick) [reply BenchPress/Generators/ResourceTypes/ApiManagementApi.cs]: All of the key names should be fixed!

Thanks for walking me through that. It took me a while, but I finally got it!

File: BenchPress/Generators/ResourceTypes/ApiManagementPolicy.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    public override IEnumerable<KeyValuePair<string, object>> GetResourceParameters(TestMetadata m)
+    {
+        return new[]
+        {
+            Param("ResourceType", "ApiManagementPolicy"),
+            Param("ResourceName", m.ResourceName),
+            Param("ResourceGroupName", m.ExtraProperties["resourceGroup"]),
+            Param("ServiceName", m.ExtraProperties["service"]),
+            Param("ApiId", m.ExtraProperties["api"])
```
Comment (msha1026) [inline BenchPress/Generators/ResourceTypes/ApiManagementPolicy.cs:21]:
```suggestion
            Param("ApiId", m.ExtraProperties["apis"])
```

**Review by msha1026 — APPROVED** [review]
Other than the one change, looks good. Thanks for adding these missing resources.

**Comment by ozboms** [comment]
Is this story also the one to update the `.PSScriptRooot` imports in `Get-ResourceByType.ps1`?

**Comment by ozboms** [comment]
> Is this story also the one to update the `.PSScriptRooot` imports in `Get-ResourceByType.ps1`?

JK just saw that Caroline is picking it up.

**Review by ozboms — CHANGES_REQUESTED** [review]
Are we missing the `.cs` class for Role Assignments?

**Comment by marclerwick** [comment]
> Are we missing the `.cs` class for Role Assignments?

It's covered in the initial comments from creating the PR:

"NOTE Due to limitations in the Generator code Key Vault Certificate and Role Assignment cannot be supported at this time. Specifically Key Vault Certificate cannot be created using bicep so the proper variable names cannot be determined for the Generator. Role Assignments have additional properties that are needed that cannot be derived from the generated ARM template."

**Review by ozboms — APPROVED** [review]
